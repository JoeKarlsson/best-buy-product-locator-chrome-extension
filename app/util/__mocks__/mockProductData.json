{
  "from": 1,
  "to": 1,
  "currentPage": 1,
  "total": 1,
  "totalPages": 1,
  "queryTime": "0.175",
  "totalTime": "0.182",
  "partial": false,
  "canonicalUrl": "/v1/products(productCode=\"UN55NU7100FXZA\")?show=inStoreAvailability,name,sku,regularPrice,salePrice,addToCartUrl,condition&sort=inStoreAvailability&format=json&apiKey=ew0faCm5P5CuVBVGbtsI6byM",
  "products": [
    {
      "inStoreAvailability": true,
      "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
      "sku": 6200125,
      "regularPrice": 649.99,
      "salePrice": 599.99,
      "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
      "condition": "New"
    }
  ]
}
