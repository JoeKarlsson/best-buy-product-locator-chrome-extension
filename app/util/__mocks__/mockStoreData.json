{
  "from": 1,
  "to": 5,
  "currentPage": 1,
  "total": 9,
  "totalPages": 2,
  "queryTime": "0.014",
  "totalTime": "0.042",
  "partial": false,
  "canonicalUrl": "/v1/stores((area(\"55423\",15)))+products(sku in(6200125))?show=products.sku,products.name,products.shortDescription,products.salePrice,products.regularPrice,products.addToCartURL,products.url,products.image,products.customerReviewCount,products.customerReviewAverage,address,address2,city,country,detailedHours,lat,location,lng,name,longName,phone,fullPostalCode,region,storeId&pageSize=5&format=json&apiKey=ew0faCm5P5CuVBVGbtsI6byM",
  "stores": [
    {
      "products": [
        {
          "sku": 6200125,
          "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
          "shortDescription": "2160p (4K) resolution with HDRSmart TVMotion Rate 120",
          "salePrice": 599.99,
          "regularPrice": 649.99,
          "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
          "url": "https://api.bestbuy.com/click/-/6200125/pdp",
          "image": "https://img.bbystatic.com/BestBuy_US/images/products/6200/6200125_sa.jpg",
          "customerReviewCount": 457,
          "customerReviewAverage": 4.7
        }
      ],
      "address": "1000 West 78th St",
      "address2": "",
      "city": "Richfield",
      "country": "US",
      "detailedHours": [
        { "day": "Sunday", "date": "2018-08-26", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-08-27", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-08-28", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-08-29", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-08-30", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-08-31", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-01", "open": "10:00", "close": "21:00" },
        { "day": "Sunday", "date": "2018-09-02", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-09-03", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-09-04", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-09-05", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-09-06", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-09-07", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-08", "open": "10:00", "close": "21:00" }
      ],
      "lat": 44.86321,
      "lng": -93.29274,
      "name": "RICHFIELD MN",
      "longName": "Richfield",
      "phone": "612-861-3917",
      "fullPostalCode": "55423-3912",
      "region": "MN",
      "storeId": 281
    },
    {
      "products": [
        {
          "sku": 6200125,
          "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
          "shortDescription": "2160p (4K) resolution with HDRSmart TVMotion Rate 120",
          "salePrice": 599.99,
          "regularPrice": 649.99,
          "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
          "url": "https://api.bestbuy.com/click/-/6200125/pdp",
          "image": "https://img.bbystatic.com/BestBuy_US/images/products/6200/6200125_sa.jpg",
          "customerReviewCount": 457,
          "customerReviewAverage": 4.7
        }
      ],
      "address": "1235 Town Centre Dr",
      "address2": "",
      "city": "Eagan",
      "country": "US",
      "detailedHours": [
        { "day": "Sunday", "date": "2018-08-26", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-08-27", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-08-28", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-08-29", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-08-30", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-08-31", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-01", "open": "10:00", "close": "21:00" },
        { "day": "Sunday", "date": "2018-09-02", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-09-03", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-09-04", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-09-05", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-09-06", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-09-07", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-08", "open": "10:00", "close": "21:00" }
      ],
      "lat": 44.83303,
      "lng": -93.15589,
      "name": "EAGAN MN",
      "longName": "Eagan",
      "phone": "651-365-0540",
      "fullPostalCode": "55123-1067",
      "region": "MN",
      "storeId": 1055
    },
    {
      "products": [
        {
          "sku": 6200125,
          "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
          "shortDescription": "2160p (4K) resolution with HDRSmart TVMotion Rate 120",
          "salePrice": 599.99,
          "regularPrice": 649.99,
          "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
          "url": "https://api.bestbuy.com/click/-/6200125/pdp",
          "image": "https://img.bbystatic.com/BestBuy_US/images/products/6200/6200125_sa.jpg",
          "customerReviewCount": 457,
          "customerReviewAverage": 4.7
        }
      ],
      "address": "11600 Leona Rd",
      "address2": "",
      "city": "Eden Prairie",
      "country": "US",
      "detailedHours": [
        { "day": "Sunday", "date": "2018-08-26", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-08-27", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-08-28", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-08-29", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-08-30", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-08-31", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-01", "open": "10:00", "close": "21:00" },
        { "day": "Sunday", "date": "2018-09-02", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-09-03", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-09-04", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-09-05", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-09-06", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-09-07", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-08", "open": "10:00", "close": "21:00" }
      ],
      "lat": 44.86091,
      "lng": -93.42473,
      "name": "EDEN PRAIRIE MN",
      "longName": "Eden Prairie",
      "phone": "952-941-5149",
      "fullPostalCode": "55344-3803",
      "region": "MN",
      "storeId": 611
    },
    {
      "products": [
        {
          "sku": 6200125,
          "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
          "shortDescription": "2160p (4K) resolution with HDRSmart TVMotion Rate 120",
          "salePrice": 599.99,
          "regularPrice": 649.99,
          "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
          "url": "https://api.bestbuy.com/click/-/6200125/pdp",
          "image": "https://img.bbystatic.com/BestBuy_US/images/products/6200/6200125_sa.jpg",
          "customerReviewCount": 457,
          "customerReviewAverage": 4.7
        }
      ],
      "address": "14141 Aldrich Ave S",
      "address2": "",
      "city": "Burnsville",
      "country": "US",
      "detailedHours": [
        { "day": "Sunday", "date": "2018-08-26", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-08-27", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-08-28", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-08-29", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-08-30", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-08-31", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-01", "open": "10:00", "close": "21:00" },
        { "day": "Sunday", "date": "2018-09-02", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-09-03", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-09-04", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-09-05", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-09-06", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-09-07", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-08", "open": "10:00", "close": "21:00" }
      ],
      "lat": 44.74802,
      "lng": -93.28584,
      "name": "BURNSVILLE MN",
      "longName": "Burnsville",
      "phone": "952-435-9379",
      "fullPostalCode": "55337-4444",
      "region": "MN",
      "storeId": 8
    },
    {
      "products": [
        {
          "sku": 6200125,
          "name": "Samsung - 55\" Class - LED - NU7100 Series - 2160p - Smart - 4K UHD TV with HDR",
          "shortDescription": "2160p (4K) resolution with HDRSmart TVMotion Rate 120",
          "salePrice": 599.99,
          "regularPrice": 649.99,
          "addToCartUrl": "https://api.bestbuy.com/click/-/6200125/cart",
          "url": "https://api.bestbuy.com/click/-/6200125/pdp",
          "image": "https://img.bbystatic.com/BestBuy_US/images/products/6200/6200125_sa.jpg",
          "customerReviewCount": 457,
          "customerReviewAverage": 4.7
        }
      ],
      "address": "8100 Old Carriage Ct",
      "address2": "",
      "city": "Shakopee",
      "country": "US",
      "detailedHours": [
        { "day": "Sunday", "date": "2018-08-26", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-08-27", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-08-28", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-08-29", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-08-30", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-08-31", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-01", "open": "10:00", "close": "21:00" },
        { "day": "Sunday", "date": "2018-09-02", "open": "10:00", "close": "20:00" },
        { "day": "Monday", "date": "2018-09-03", "open": "10:00", "close": "21:00" },
        { "day": "Tuesday", "date": "2018-09-04", "open": "10:00", "close": "21:00" },
        { "day": "Wednesday", "date": "2018-09-05", "open": "10:00", "close": "21:00" },
        { "day": "Thursday", "date": "2018-09-06", "open": "10:00", "close": "21:00" },
        { "day": "Friday", "date": "2018-09-07", "open": "10:00", "close": "21:00" },
        { "day": "Saturday", "date": "2018-09-08", "open": "10:00", "close": "21:00" }
      ],
      "lat": 44.77825,
      "lng": -93.41347,
      "name": "SHAKOPEE MN",
      "longName": "Shakopee",
      "phone": "952-233-0387",
      "fullPostalCode": "55379-3160",
      "region": "MN",
      "storeId": 1463
    }
  ],
  "warnings": "Distances are in terms of miles (default)"
}
