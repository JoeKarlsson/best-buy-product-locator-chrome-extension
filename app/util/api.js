import { DEFAULT_ERROR_MESSAGE } from '../constants/constants';

const isNotHappyStatus = status => status >= 400;

const handleError = (err) => {
  const errMsg = err.toString();
  // eslint-disable-next-line no-console
  console.error(errMsg);
  return {
    errMsg,
    msg: DEFAULT_ERROR_MESSAGE,
  };
};

const handleResponse = (response) => {
  const { status } = response;
  console.log('response', response);
  if (isNotHappyStatus(status)) {
    return handleError('Bad response from server');
  }
  return response.json();
};

const api = (url, options) => fetch(url, options)
  .then(handleResponse)
  .catch(handleError);

export default api;
