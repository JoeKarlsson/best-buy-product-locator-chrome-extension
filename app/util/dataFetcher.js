import handleError from './handleError';

export default function fetchData(url) {
  return new Promise(async (resolve, reject) => {
    try {
      const response = await fetch(url);

      if (response.ok) {
        resolve(await response.json());
      }
    } catch (err) {
      const errMsg = handleError(err);
      reject(errMsg);
    }
  });
}
