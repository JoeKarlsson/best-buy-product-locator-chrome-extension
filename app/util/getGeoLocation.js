const getGeoLocation = () => {
  const options = {
    enableHighAccuracy: true,
    maximumAge: 10000,
    timeout: 7000,
  };

  return new Promise((ok, err) => {
    navigator.geolocation.getCurrentPosition(
      (position) => {
        if (position.coords) {
          ok(position);
        }
      },
      (errMsg) => {
        err(errMsg);
      },
      options,
    );
  });
};

export default getGeoLocation;
