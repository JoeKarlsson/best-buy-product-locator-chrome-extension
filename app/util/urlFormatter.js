import { BASE_URL } from '../constants/constants';

export const constructProductURL = modelNumber => `${BASE_URL}products(modelNumber=${modelNumber})?apiKey=${
  process.env.BBY_API_KEY
}&sort=inStoreAvailability.asc&show=inStoreAvailability,name,sku,regularPrice,salePrice,addToCartUrl,condition&format=json`;

export const constructStoreURL = (skuId, zipCode, miles) => `${BASE_URL}stores((area(${zipCode},${miles})))+products(sku%20in%20(${skuId}))?apiKey=${
  process.env.BBY_API_KEY
}&show=products.sku,products.name,products.shortDescription,products.salePrice,products.regularPrice,products.addToCartURL,products.url,products.image,products.customerReviewCount,products.customerReviewAverage,address,address2,city,country,detailedHours,lat,location,lng,name,longName,phone,fullPostalCode,region,storeId&pageSize=5&format=json`;

export const constructNearestStoreMapUrl = (city, region, lat, lng) => `https://www.google.com/maps/search/Best+Buy+${city}+${region}/@${lat},${lng}`;
