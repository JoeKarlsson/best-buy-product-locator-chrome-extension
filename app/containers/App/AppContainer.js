import React, { Component } from 'react';
import App from './App';
import {
  constructProductURL,
  constructStoreURL,
  constructNearestStoreMapUrl,
} from '../../util/urlFormatter';
import api from '../../util/api';

class AppContainer extends Component {
  static isValidProductData(productData) {
    return productData.products && productData.products.length > 0;
  }

  static isValidStoreData(storeData) {
    return storeData.stores && storeData.stores.length > 0;
  }

  constructor() {
    super();

    this.state = {
      addToCartUrl: '',
      nearestStore: '',
      nearestStoreMapUrl: '',
      price: 0,
      isError: false,
      isLoading: true,
    };
  }

  async componentDidMount() {
    // check local storage for existing model numbers on page
    chrome.storage.local.get(['modelNumbers'], (result) => {
      if (result.modelNumbers && result.modelNumbers.length > 0) {
        result.modelNumbers.forEach(async (modelNumber) => {
          this.getAvailability(modelNumber);
        });
      }
    });
    // set up listener for model number DOM scraper
    chrome.runtime.onMessage.addListener((modelNumbers) => {
      modelNumbers.forEach(async (modelNumber) => {
        this.getAvailability(modelNumber);
      });
    });
  }

  getAvailability = async (modelNumber) => {
    const productURL = constructProductURL(modelNumber);
    const productData = await api(productURL);

    if (AppContainer.isValidProductData(productData)) {
      const product = productData.products[0];
      const skuId = product.sku;

      // TODO: figure out how to get zipcode or geolocation coordinates
      const storeURL = constructStoreURL(skuId, '55423', '15');
      const storeData = await api(storeURL);

      if (AppContainer.isValidStoreData(storeData)) {
        const nearestStore = storeData.stores[0];

        const {
          city, region, lat, lng
        } = nearestStore;

        this.setState({
          addToCartUrl: product.addToCartUrl,
          nearestStore: `${city}, ${region}`,
          nearestStoreMapUrl: constructNearestStoreMapUrl(city, region, lat, lng),
          price: product.salePrice || product.regularPrice,
          isLoading: false,
        });
      }
    } else {
      this.setState({
        isError: true,
        isLoading: false,
      });
    }
  };

  render() {
    const {
      nearestStore,
      nearestStoreMapUrl,
      price,
      addToCartUrl,
      isLoading,
      isError,
    } = this.state;

    return (
      <App
        nearestStore={nearestStore}
        nearestStoreMapUrl={nearestStoreMapUrl}
        price={price}
        addToCartUrl={addToCartUrl}
        isLoading={isLoading}
        isError={isError}
      />
    );
  }
}

export default AppContainer;
