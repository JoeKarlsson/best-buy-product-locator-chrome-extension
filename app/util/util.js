import { BASE_URL } from '../constants/constants';
import dataFetcher from './dataFetcher';

const constructProductURL = (baseURL, modelNumber) => `${BASE_URL}products(modelNumber=${modelNumber})?apiKey=${
  process.env.BBY_API_KEY
}&sort=inStoreAvailability.asc&show=inStoreAvailability,name,sku,regularPrice,salePrice,addToCartUrl,condition&format=json`;

const constructStoreURL = (baseURL, skuId, zipCode, miles) => `${BASE_URL}stores((area(${zipCode},${miles})))+products(sku%20in%20(${skuId}))?apiKey=${
  process.env.BBY_API_KEY
}&show=products.sku,products.name,products.shortDescription,products.salePrice,products.regularPrice,products.addToCartURL,products.url,products.image,products.customerReviewCount,products.customerReviewAverage,address,address2,city,country,detailedHours,lat,location,lng,name,longName,phone,fullPostalCode,region,storeId&pageSize=5&format=json`;

export function getModelNumber() {
  // eslint-disable-next-line no-unused-vars
  const modelRegex = new RegExp(/\b((?=[A-Za-z/ -]{0,19}\d)[A-Za-z0-9/ -]{4,20})\b/);
  // TODO: write DOM-scraper function to get product model number
}

export const fetchProductData = async (modelNumber) => {
  try {
    return dataFetcher(constructProductURL(BASE_URL, modelNumber));
  } catch (err) {
    return err;
  }
};

export function fetchStoreData(skuId, zipCode, miles) {
  try {
    return dataFetcher(constructStoreURL(BASE_URL, skuId, zipCode, miles));
  } catch (err) {
    return err;
  }
}
