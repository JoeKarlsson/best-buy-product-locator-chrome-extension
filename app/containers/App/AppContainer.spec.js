import React from 'react';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock';
import AppContainer from './App';
import { constructProductURL, constructStoreURL } from '../../util/urlFormatter';
import mockProductData from '../../util/__mocks__/mockProductData.json';
import mockStoreData from '../../util/__mocks__/mockStoreData.json';

describe('AppContainer', () => {
  beforeEach(() => {
    fetchMock.reset();
  });

  describe('rendering', () => {
    describe('initial state', () => {
      it('is rendered heathy state', () => {
        const modelNumber = 'UN55NU7100FXZA';
        const productUrl = constructProductURL(modelNumber);
        fetchMock.once(productUrl, mockProductData);

        const zipCode = '55126';
        const miles = 5;
        const skuId = '55126';
        const storeUrl = constructStoreURL(skuId, zipCode, miles);
        fetchMock.once(storeUrl, mockStoreData);

        const component = renderer.create(<AppContainer />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
