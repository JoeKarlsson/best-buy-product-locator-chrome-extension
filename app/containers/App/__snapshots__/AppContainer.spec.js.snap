// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer rendering initial state is rendered heathy state 1`] = `null`;
